cpp/src/arrow/dataset/file_base.cc

```diff
@@ -89,6 +89,28 @@ Result<std::shared_ptr<io::InputStream>> FileSource::OpenCompressed(
   return io::CompressedInputStream::Make(codec.get(), std::move(file));
 }
 
+Result<std::shared_ptr<io::InputStream>> FileSource::OpenRange(int64_t start,
+                                                               int64_t end) const {
```

Comment:
Does this mean the input file will be truncated exactly from `start` to `end` then deliver to fragment?

In some cases fuzzed match might be required. E.g. Spark picks parquet row groups whose binary midpoints are located in [start, end).

Reply:
The implementation behind the interface should be changeable depending on the file format. This draft PR is more of a proposal for how the interface for specifying these options should be. I think it should also work for your parquet use case. Maybe not?

Reply:
I see. Then the codes may also work for my case based on current shape of the API. My speculation is to have a new C++ function to split a single `ParquetFileFragment` to multiple.

Reply:
I understand your use case to be: you have one Parquet file, you want to specify a byte range, and you want to read all the row groups that fit in that byte range. 

I think with this API here, you don't need to split a single ParquetFileFragment into multiple, nor do you need a new C++ function. We can just incorporate your changes in file_parquet.cc and have the ParquetFileFragment interpret the start and end bytes in the way that you specified instead of insisting that they align on proper row group boundaries.

Reply:
OK. Then it seems to me workable to have the fragments produced from dataset factory directly without re-splitting. 

And a naive question around the codes: should we avoid `set_bounds` since it brings extra states to file fragment class?

Reply:
I think my idea is to have some state in the file fragment class that can be interpreted differently depending on the actual file fragment format. Other people might have different thoughts?

Reply:
if you are on board with this API @zhztheplayer  I can probably incorporate your code into my PR. Or I can try to get my PR thru first and then you can maybe adapt your PR to my API.